Rewrite `AtomicTask`, fixing unsoundness

carllerche opened the pull request:
**This is a work in progress** I am creating the PR early in case anyone wants to take an early look.

The `AtomicTask` abstraction was initially added as an internal utility for a specific use case. It has since been included in the public API.

The original implementation was edited incrementally, unfortunately adding bugs. An attempt (#868) was made to fix an unsoundness issue as well as another bug, but this unfortunately introduced new unsoundness due to another pre-existing bug that just happened to be hidden by the bug that was fixed... i.e., it was a mess.

This PR attempts to resolve the issues by taking starting from scratch.

Preliminary testing finds that this PR fixes the issue tracked in https://github.com/tokio-rs/tokio/issues/236, but more thorough testing and reviewing is needed.

### Remaining

* [x] More extensive testing against Tokio
* [x] Document implementation.

cramertj commented on src/task_impl/atomic_task.rs:
```diff
@@ -86,90 +73,96 @@ impl AtomicTask {
+                                // The atomic exchange was successful, now break
+                                // out of the loop os that task stored in
+                                // `notify` can be notified (if set).
+                                break;
```
Nit: this loop is a bit hard for me to read w/ this `break` in it. Can you inline the `if let` below into the this branch and put a `return` here?

cramertj commented on src/task_impl/atomic_task.rs:
```diff
@@ -86,90 +73,96 @@ impl AtomicTask {
                 task.notify();
             }
+            _ => {
+                // TODO: Assert valid state
```
In this branch, two people called register at ~ the same time, so it doesn't matter that we're technically the *second* one to call register if we reach this case?

carllerche replied:
Basically... concurrently calling `register` doesn't really make sense. `AtomicTask` only tries to maintain memory safety, but the `register` is lost.

cramertj commented on src/task_impl/atomic_task.rs:
```diff
@@ -86,90 +73,96 @@ impl AtomicTask {
-                curr = actual;
-
+            }
+            _ => {
```
Nit: can you list out the remaining cases, just to make it clear that all of them should indeed be no-ops?

(REGISTERING, NOTIFYING, REGISTERING | NOTIFYING)

cramertj commented on src/task_impl/atomic_task.rs:
```diff
@@ -86,90 +73,96 @@ impl AtomicTask {
+                            Err(actual) => {
+                                // Update `curr` for the next iteration of the
+                                // loop
+                                curr = actual;
```
`actual` should be only `REGISTERING | NOTIFYING`, right? Can you update the code to specifically handle this case, rather than matching on any value of `actual`? (and panic or something in other cases)

cramertj commented on src/task_impl/atomic_task.rs:
```diff
@@ -86,90 +73,96 @@ impl AtomicTask {
-                // Currently in a read locked state, this implies that `notify`
-                // is currently being called on the old task handle. So, we call
-                // notify on the new task handle
+            NOTIFYING => {
```
Shouldn't it also call `task.notify()` if we read `REGISTERING | NOTIFYING`?

carllerche replied:
I'm treating `REGISTERING` being set when calling `register` as a caller bug. If you want, I can add a `panic` in this case?

cramertj commented on src/task_impl/atomic_task.rs:
```diff
@@ -86,90 +73,96 @@ impl AtomicTask {
+
+                    loop {
+                        let res = self.state.compare_exchange(
+                            curr, WAITING, AcqRel, Acquire);
```
Should this be `Release, Acquire`? On success, we just need the previous writes to `self.task` to be written. On failure, we only need to be able to read others' writes to `self.task`.

carllerche replied:
It has to be `AcqRel` as the success ordering must be stronger than the failure ordering (i.e. a super set).

cramertj replied:
TIL! Thanks.

cramertj commented:
Thanks for doing this! This code was in some definite need of TLC. Overall, it would help me be more confident about the code if you explicitly enumerated all of the possible cases and state transitions with every atomic op. Exhaustive matching makes it much easier for me to reason about.

carllerche commented:
I addressed the feedback (or replied to your comment). I kept the state assertions as debug asserts, but if you think it would be better to perform the checks at runtime, I can switch it.

cramertj commented on src/task_impl/atomic_task.rs:
```diff
@@ -86,39 +73,75 @@ impl AtomicTask {
+                                // concurrent thread called `notify`. In this
+                                // case, `actual` **must** be `REGISTERING |
+                                // `NOTIFYING`.
+                                debug_assert_eq!(curr, REGISTERING | NOTIFYING);
```
Should this be comparing `actual`, rather than `curr`?

carllerche replied:
ugh yeah, I moved that assert up (from after curr was set to actual).

cramertj commented:
LGTM modulo the one nit about the `assert`.

carllerche commented:
I pushed docs. I'm going to try to do some more extensive testing against Tokio now.

cramertj commented:
SGTM-- thanks again for doing this. Let me know when you're comfortable with it and I'll merge.

carllerche commented:
Ok, as far as I can tell, everything works.

carllerche commented:
Looks like CI is passing!

cramertj commented:
@carllerche To confirm: you'll follow up w/ a matching change to 0.2?

carllerche commented:
@cramertj yeah, I can work on that right after I am done w/ https://github.com/tokio-rs/tokio/pull/243

cramertj commented:
@carllerche Great, thanks!
